docs/design/datacontracts/data_descriptor.md

```diff
@@ -0,0 +1,310 @@
+
+* `"name": "type name"` the name of each type
+* optional `"size": int | "indeterminate"` if omitted the size is indeterminate
+* optional `"fields": FIELD_ARRAY` if omitted same as a field array of length zero
+
+Each `FIELD_ARRAY` is an array of dictionaries each containing keys:
+
+* `"name": "field name"` the name of each field
+* `"type": "type name"` the name of a primitive type or another type defined in the same /logical/ descriptor
+* optional `"offset": int | "unknown"` the offset of the field or "unknown". If omitted, same as "unknown".
+
+Note that the logical descriptor does not contain "unknown" offsets.
+
+Rationale: "unknown" offsets may be used to document in the physical JSON descriptor that another
+physical descriptor in the "baseline" graph is expected to provide the offset of the field.
+
+### Global values
+
+The global values will be in an array, with each value described by a dictionary containing keys:
+
+* `"name": "global value name"` the name of the global value
+* `"type": "type name"` the type of the global value
+* optional `"value": VALUE | "unknown"` the value of the global value or "unknown". If omitted, same as "unknown".
+
+Note that the logical descriptor does not contain "unknown" values.
+
+The `VALUE` may be a JSON numeric constant integer or a string containing a signed or unsigned
+decimal or hex (with prefix `0x` or `0X`) integer constant.  The constant must be within the range
+of the type of the global value.
+
+For pointer and nuint globals, the value may be assumed to fit in a 64-bit unsigned integer.  For
+nint globals, the value may be assumed to fit in a 64-bit signed integer.
+
+If the value is specified as "unknown" another physical descriptor in the "baseline" graph is
+expected to provide the value.
+
+## Physical binary blob descriptor
+
+### Version
+
+This is version 0 of the physical binary blob format.
+
+### Summary
+
+The binary blob format for a physical descriptor is expected to be stored in the memory space of a
+target .NET runtime and is thus likely to be the final descriptor in the "baseline" graph.
+
+It is likely that the physical descriptor will be a part of a build-time constant in the disk image
+of a .NET runtime, and the format is designed to be compact and specifiable as a (likely
+machine-generated) compile-time constant in a suitable source language.
+
+The data descriptor forms one part of an overall physical data contract descriptor in a targret .NET
+runtime and as such this format does not specify a "magic number", or "well known symbol", or
+another means of identifying the blob within a target process.  Additionally the version of the
+binary blob data descriptor is expected to be stored within the larger enclosing data contract
+descriptor and is not included here.
+
+### Blob
+
+Multi-byte values are in the target platform endianness.
+
+The format is:
+
+```c
+struct BinaryBlobDataDescriptor
+{
+    struct Directory {
+        uint32_t TypesStart;
```

Comment:
So this comes from teh requirement that we would like to be able to read binary descriptors directly out of object files without understanding the object file format.   That means that if the object file is produced by a C compiler, we want to be able to ignore any padding or alignment that the C compiler adds.  So because we have fixed-size arrays of structs as part of `BinaryBlobDataDescriptor` I think the C compiler may add padding between when one array ends and the next one starts.  So to make the blob self-describing, we need to capture the offsets of the fields of `BinaryBlobDataDescriptor` as part of itself.

If the C compiler adds padding.

Reply:
Does that mean this field is the relative offset from the start of the BinaryBlobDescriptor to the start of the Types array and we'd never expect a reader to know that BinaryBlobDescriptor has a Types field? If I followed you correctly I'd be tempted to omit all of those array fields from the struct definition or add a comment saying those fields are implementation details that a reader shouldn't rely upon.

Reply:
> I'd be tempted to omit all of those array fields from the struct definition or add a comment saying those fields are implementation details that a reader shouldn't rely upon.

+1. We do not need to spec the intermediate artifacts used in the process of building the final descriptor. They can be just our internal implementation detail.

Reply:
@jkotas ok, I'm gonna take one more pass at simplifying things:

1. only document the json physical format (but with a second flavor for the in-proc version that has an aux array for the global pointers)
2. keep the "object file blob" stuff as an implementation detail for tooling that will create the in-proc json data descriptor.
3. Drop the requirements to be able to scrape the on-disk representation out of the spec.

The build tooling (I'm thinking of it as `cdac-build-tool`) will work like this:
* runtime build generates one or more object files with blobs containing offsets and struct sizes
* `cdac-build-tool` scrapes the object files and a baseline and writes out a C file that contains a json data descriptor and an aux array of globals
* we compile the C file and link it into the runtime

At diagnostic time:
* DAC tools find the data descriptor in target process memory (not on disk)